text/1002-required-by-default.rst

```diff
@@ -0,0 +1,178 @@
+..
+    Status: Draft
+    Type: Guideline
+    Created: 2020-06-17
+    RFC PR: `edgedb/rfcs#0001 <https://github.com/edgedb/rfcs/pull/1>`_
+
+===============================================================
+RFC 1002: REQUIRED qualifier by default in properties and links
+===============================================================
+
+This RFC describes the reasons why DDL and SDL declarations for properties
+and links assume non-optionality by default.
+
+
+Motivation
+==========
+
+The data definition language, and more importantly, the schema definition
+language, which users use to communicate with EdgeDB, need to avoid
+unnecessary verbosity.
+
+This requires some conventions and assumptions about what the user expects
+most of the time. Those assumptions shouldn't bar the user from changing
+their mind when the need arises.
+
+A default value has far reaching consequences which we cover in this
+document.
+
+
+Implementation
+==============
+
+Since the discussion always applies to properties and links, we're referring
+to them simply as *fields* below.
+
+Assume REQUIRED qualifiers when no explicit qualifier is provided
+-----------------------------------------------------------------
+
+The main advantage of assuming that all fields are required unless otherwise
+specified is its resilience to migration into optionality later. Client code
+that was forced to provide values before will continue working and there are
+no incompatible empty sets in the database prior to the migration.
```

Comment:
That's not quite true. Let's say REQUIRED is the default and you had a schema:

```
type User {
  property name -> str;
  property email -> str;
}
```

and you had a query like this:

```
SELECT (User.name, User.email)
```

that was used to return ALL emails. Note that we can come up with many other queries; this one while impractical still demonstrates the problem.

**Then** you migrate the schema to 

```
type User {
  optional property name -> str;
  property email -> str;
}
```

Now your `SELECT (User.name, User.email)` query works differently. It would omit user emails for users that don't have a name.

So on the surface it is easier to migrate from "required" to "optional", but in practice some existing queries (especially analytical) would in fact break and start returning incorrect results.

On the other hand, migrating from "optional" to "required" would require you to specify a "default".  But then all of the existing queries in your application would continue to work correctly.

Therefore I'm really not sure that the flipping current behavior would be an improvement, quite the opposite.

Reply:
> Therefore I'm really not sure that the flipping current behavior would be an improvement, quite the opposite.

Making existing correct queries silently incorrect is indeed a big problem. That, along with the migration, might be enough to reject this entire proposal.

My intuition is that most fields will stay at whatever the *default* is. With the default being `OPTIONAL` it's more likely that people will write queries that unintentionally omit objects with missing pointers. And that's without any migrations, just bad queries from the get go.

You have the most experience with this. How do you write queries and computables? Are you doing defensive programming against empty sets? Do you come back to the schema to ensure pointers are required?

Reply:
I agree with Yury. The semantics of empty set make `required` -> `optional` change a potentially much sneakier and nastier thing.  When you make a pointer `required`, you will get explicit errors from the database asking you to provide missing data at the time of the migration, and either set a default on a pointer, or change your `INSERT`s (which would start to fail explicitly).  So I don't really see a significant *resilience* advantage to migration into optionality as opposed to vice-versa.

Another problem with making `required` the default lies with link properties, which are always optional by design.

Reply:
> My intuition is that most fields will stay at whatever the default is. With the default being OPTIONAL it's more likely that people will write queries that unintentionally omit objects with missing pointers. And that's without any migrations, just bad queries from the get go.

I assume that people will mostly write queries that selects simple shapes of data (like GraphQL), where dealing with empty sets isn't often necessary or painful.

Analytical queries will be presumably written in a way to account for empty sets, and that's why the current default to "optional" makes sense for them (flipping to "required" is easier).

> You have the most experience with this. 

That would be @elprans and @vpetrovykh :)

> How do you write queries and computables? 

There is a coalesce operator `??` and `?=` to deal with empty sets. The rules of using them aren't extremely intuitive though.

> Are you doing defensive programming against empty sets?

I don't think we need to advocate for "defensive programming", but we do need to have an easy guide on how to best handle empty sets in different scenarios. I don't think we have such a high-level document yet.

> Do you come back to the schema to ensure pointers are required?

Ideally yes, you need to know your schema when writing queries against it.

Reply:
> might be enough to reject this entire proposal.

Let's not reject it, but update this RFC to be "Informational" and explain why we have "optional" as default.

Reply:
> My intuition is that most fields will stay at whatever the default is. With the default being OPTIONAL it's more likely that people will write queries that unintentionally omit objects with missing pointers. And that's without any migrations, just bad queries from the get go.

This isn't different from SQL or any other database that allows (and defaults to) NULLs, which is to say most of them. With that in mind, I think that most people, who have some familiarity with databases, would be aware of the problem, and the best we can do is improve the ergonomics around handling (and guarding against) missing data.  The coalescing operators are one thing, but there's also the planned addition of the `<REQUIRED _>` cast, which can be used to guard against empty sets in queries where they're really not desired.

Reply:
I disagree that "tuple case" is the main counter-argument against required by default. I think that case should be dealt with regardless of this default.

The issue with a `link` is more important. But I think it's fine if `link` will be required by default and `multi` link will be optional by default.

Reply:
> But I think it's fine if link will be required by default and multi link will be optional by default.

I find that would be rather unintuitive.

Reply:
I also think that the dissonance between the requiredness of `single` and `multi` is a larger issue than inconsistency with query args and is an argument *against* the original proposal.